source/common/http/http2/codec_impl.cc

```diff
@@ -506,24 +506,34 @@ void ConnectionImpl::sendPendingFrames() {
   }
 }
 
-void ConnectionImpl::sendSettings(uint64_t codec_options) {
+void ConnectionImpl::sendSettings(const Http2Settings& http2_settings) {
+  ASSERT(http2_settings.hpack_table_size_ <= Http2Settings::MAX_HPACK_TABLE_SIZE);
+  ASSERT(Http2Settings::MIN_MAX_CONCURRENT_STREAMS <= http2_settings.max_concurrent_streams_ &&
+         http2_settings.max_concurrent_streams_ <= Http2Settings::MAX_MAX_CONCURRENT_STREAMS);
+  ASSERT(Http2Settings::MIN_INITIAL_WINDOW_SIZE <= http2_settings.initial_window_size_ &&
+         http2_settings.initial_window_size_ <= Http2Settings::MAX_INITIAL_WINDOW_SIZE);
+
   std::vector<nghttp2_settings_entry> iv = {
-      {NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, MAX_CONCURRENT_STREAMS},
-      {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, DEFAULT_WINDOW_SIZE}};
+      {NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS, http2_settings.max_concurrent_streams_},
+      {NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE, http2_settings.initial_window_size_}};
 
-  if (codec_options & CodecOptions::NoCompression) {
-    iv.push_back({NGHTTP2_SETTINGS_HEADER_TABLE_SIZE, 0});
-    conn_log_debug("disabling header compression", connection_);
+  if (http2_settings.hpack_table_size_ != NGHTTP2_DEFAULT_HEADER_TABLE_SIZE) {
+    iv.push_back({NGHTTP2_SETTINGS_HEADER_TABLE_SIZE, http2_settings.hpack_table_size_});
+    conn_log_debug("setting HPACK table size to {}", connection_, http2_settings.hpack_table_size_);
   }
 
   int rc = nghttp2_submit_settings(session_, NGHTTP2_FLAG_NONE, &iv[0], iv.size());
   ASSERT(rc == 0);
   UNREFERENCED_PARAMETER(rc);
 
   // Increase connection window size up to our default size.
-  rc = nghttp2_submit_window_update(session_, NGHTTP2_FLAG_NONE, 0,
-                                    DEFAULT_WINDOW_SIZE - NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE);
-  ASSERT(rc == 0);
+  ASSERT(http2_settings.initial_window_size_ >= NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE);
+  if (http2_settings.initial_window_size_ > NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE) {
```

Comment:
i made http2_settings.initial_window_size_ >= NGHTTP2_INITIAL_CONNECTION_WINDOW_SIZE
in config schemas, so i just use > to compare them.

i guess they are different configs,  so it make sense to not mess them togetther.
i removed the assert and use not equal now.